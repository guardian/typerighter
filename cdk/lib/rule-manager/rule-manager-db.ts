import { Port, SecurityGroup, SubnetType } from "@aws-cdk/aws-ec2";
import {
  Credentials,
  DatabaseInstanceEngine,
  PostgresEngineVersion,
  StorageType,
  SubnetGroup,
} from "@aws-cdk/aws-rds";
import type { App } from "@aws-cdk/core";
import { Duration, RemovalPolicy, SecretValue } from "@aws-cdk/core";
import { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import { GuParameter, GuStack } from "@guardian/cdk/lib/constructs/core";
import { GuSecurityGroup, GuVpc } from "@guardian/cdk/lib/constructs/ec2";
import { GuDatabaseInstance } from "@guardian/cdk/lib/constructs/rds";
export class RuleManagerDB extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const dbPort = 5432;

    const app = "rule-manager-db";

    const parameters = {
      VpcId: new GuParameter(this, "VPC", {
        type: "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
        description: "Virtual Private Cloud to run EC2 instances within",
        default: "/account/vpc/default/id",
      }),
      MasterDBUsername: new GuParameter(this, "MasterDBUsername", {
        description: "Master DB username",
        default: "rule_manager",
        type: "String",
      }),
      PrivateVpcSubnets: new GuParameter(this, "PrivateSubnets", {
        type: "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
        description: "Subnets to run the ASG and instances within",
        default: "/account/vpc/default/private.subnets",
      }),
      AccessSecurityGroupID: new GuParameter(this, "AccessSecurityGroupID", {
        description:
          "Id of the security group from which access to the DB will be allowed",
        type: "AWS::EC2::SecurityGroup::Id",
      }),
    };

    /* Resources */

    const vpc = GuVpc.fromId(this, "Vpc", {
      vpcId: parameters.VpcId.valueAsString,
    });

    const subnets = GuVpc.subnets(
      this,
      parameters.PrivateVpcSubnets.valueAsList
    );

    const dbSecurityGroup = new GuSecurityGroup(this, "DBSecurityGroup", {
      app,
      description: "DB security group servers",
      vpc,
      allowAllOutbound: true,
    });

    dbSecurityGroup.connections.allowFrom(
      SecurityGroup.fromSecurityGroupId(
        this,
        "AccessSecurityGroup",
        parameters.AccessSecurityGroupID.valueAsString
      ),
      Port.tcp(dbPort)
    );

    new GuDatabaseInstance(this, "RuleManagerRDS", {
      app,
      vpc,
      vpcSubnets: { subnetType: SubnetType.PRIVATE },
      allocatedStorage: 50,
      allowMajorVersionUpgrade: true,
      autoMinorVersionUpgrade: true,
      backupRetention: Duration.days(10),
      engine: DatabaseInstanceEngine.postgres({
        version: PostgresEngineVersion.VER_11,
      }),
      instanceType: "t3.micro",
      instanceIdentifier: `typerighter-rule-manager-db-${this.stage}`,
      parameters: {
        max_connections: "100",
        maintenance_work_mem: "245760",
        work_mem: "38912",
        checkpoint_completion_target: "0.9",
        shared_buffers: "131072",
        synchronous_commit: "off",
        effective_cache_size: "393216",
        random_page_cost: "1.1",
      },
      subnetGroup: new SubnetGroup(this, "DBSubnetGroup", {
        vpc,
        vpcSubnets: { subnets },
        description: "Private subnet for typerighter rule-manager database",
      }),
      credentials: Credentials.fromPassword(
        parameters.MasterDBUsername.valueAsString,
        SecretValue.ssmSecure(
          `/${this.stage}/${this.stack}/typerighter-rule-manager/db.default.password`,
          "1"
        )
      ),
      multiAz: true,
      port: dbPort,
      preferredBackupWindow: "02:00-02:30",
      preferredMaintenanceWindow: "Mon:06:30-Mon:07:00",
      securityGroups: [dbSecurityGroup],
      storageEncrypted: true,
      storageType: StorageType.GP2,
      removalPolicy: RemovalPolicy.SNAPSHOT,
    });
  }
}
