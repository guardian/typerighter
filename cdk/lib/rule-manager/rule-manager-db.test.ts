import "@aws-cdk/assert/jest";
import { SynthUtils } from "@aws-cdk/assert";
import { App } from "@aws-cdk/core";
import { RuleManagerDB } from "./rule-manager-db";

describe("The rule manager db stack", () => {
    it("matches the snapshot", () => {
        const app = new App();
        const stack = new RuleManagerDB(app, "rule-manager-db", {
            stack: "flexible",
        });

        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
    })
})
