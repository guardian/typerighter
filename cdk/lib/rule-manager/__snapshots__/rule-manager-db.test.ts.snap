// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The rule manager db stack matches the snapshot 1`] = `
Object {
  "Parameters": Object {
    "AccessSecurityGroupID": Object {
      "Description": "Id of the security group from which access to the DB will be allowed",
      "Type": "AWS::EC2::SecurityGroup::Id",
    },
    "MasterDBUsername": Object {
      "Default": "rule_manager",
      "Description": "Master DB username",
      "Type": "String",
    },
    "PrivateSubnets": Object {
      "Default": "/account/vpc/default/private.subnets",
      "Description": "Subnets to run the ASG and instances within",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
    "VPC": Object {
      "Default": "/account/vpc/default/id",
      "Description": "Virtual Private Cloud to run EC2 instances within",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
  },
  "Resources": Object {
    "DBSecurityGroupRulemanagerdb3E93B6BD": Object {
      "Properties": Object {
        "GroupDescription": "DB security group servers",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "rule-manager-db",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/typerighter",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": Object {
          "Ref": "VPC",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DBSecurityGroupRulemanagerdbfromrulemanagerdbAccessSecurityGroup1C2D2DCD5432F746BBF3": Object {
      "Properties": Object {
        "Description": "from rulemanagerdbAccessSecurityGroup1C2D2DCD:5432",
        "FromPort": 5432,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "DBSecurityGroupRulemanagerdb3E93B6BD",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Ref": "AccessSecurityGroupID",
        },
        "ToPort": 5432,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "DBSubnetGroup": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Private subnet for typerighter rule-manager database",
        "SubnetIds": Object {
          "Ref": "PrivateSubnets",
        },
        "Tags": Array [
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/typerighter",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
    "RuleManagerRDSEBF20BD9": Object {
      "DeletionPolicy": "Snapshot",
      "Properties": Object {
        "AllocatedStorage": "50",
        "AllowMajorVersionUpgrade": true,
        "AutoMinorVersionUpgrade": true,
        "BackupRetentionPeriod": 10,
        "CopyTagsToSnapshot": true,
        "DBInstanceClass": "db.t3.micro",
        "DBInstanceIdentifier": Object {
          "Fn::Join": Array [
            "",
            Array [
              "typerighter-rule-manager-db-",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "DBParameterGroupName": Object {
          "Ref": "RuleManagerRDSRDSParameterGroupE6969C6C",
        },
        "DBSubnetGroupName": Object {
          "Ref": "DBSubnetGroup",
        },
        "DeletionProtection": true,
        "Engine": "postgres",
        "EngineVersion": "11",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:ssm-secure:/",
              Object {
                "Ref": "Stage",
              },
              "/flexible/typerighter-rule-manager/db.default.password:1}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Ref": "MasterDBUsername",
        },
        "MultiAZ": true,
        "Port": "5432",
        "PreferredBackupWindow": "02:00-02:30",
        "PreferredMaintenanceWindow": "Mon:06:30-Mon:07:00",
        "PubliclyAccessible": false,
        "StorageEncrypted": true,
        "StorageType": "gp2",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "rule-manager-db",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/typerighter",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VPCSecurityGroups": Array [
          Object {
            "Fn::GetAtt": Array [
              "DBSecurityGroupRulemanagerdb3E93B6BD",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBInstance",
      "UpdateReplacePolicy": "Snapshot",
    },
    "RuleManagerRDSRDSParameterGroupE6969C6C": Object {
      "Properties": Object {
        "Description": "Parameter group for postgres11",
        "Family": "postgres11",
        "Parameters": Object {
          "checkpoint_completion_target": "0.9",
          "effective_cache_size": "393216",
          "maintenance_work_mem": "245760",
          "max_connections": "100",
          "random_page_cost": "1.1",
          "shared_buffers": "131072",
          "synchronous_commit": "off",
          "work_mem": "38912",
        },
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "rule-manager-db",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/typerighter",
          },
          Object {
            "Key": "Stack",
            "Value": "flexible",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::RDS::DBParameterGroup",
    },
  },
}
`;
